frontend_lib/src/helper.js

```diff
@@ -396,71 +395,62 @@ export const getCurrentContentVersionNumber = (appFeatureMode, content, timeline
   return timeline.filter(t => t.timelineType === 'revision' && t.hasBeenRead).length
 }
 
-// WIP
 export const wrapMentionsInSpanTags = (text) => {
   try {
     const mentionRegex = /(^|\s)@([a-zA-Z0-9\-_]+)($|\s)/
 
     const parser = new DOMParser()
     const parsedText = parser.parseFromString(text, 'text/html')
 
-    const depthFirstSearchAndMentionAnalysis = node => {
-      node.forEach(childNode => {
-        if(childNode.nodeName === '#text' && childNode.textContent.includes('@')) {
-          const mentionsInThisChild = childNode.textContent.split(/\s/).filter(token => mentionRegex.test(token))
-          console.log(mentionsInThisChild)
-        } else if (!(childNode.nodeName === 'SPAN' && childNode.className === 'mention')) {
-          if(childNode.childNodes.length > 0) depthFirstSearchAndMentionAnalysis(childNode.childNodes)
+    const depthFirstSearchAndMentionAnalysis = childNodesList => {
+      let childNodesListCopy = [...childNodesList]
+      let i = 0
+
+      childNodesListCopy.forEach((node) => {
+        let value = node.nodeValue
+
+        if(node.nodeName === '#text' && value.includes('@')) {
+          const mentionsInThisNode = value.split(/\s/).filter(token => mentionRegex.test(token))
+
+          if (mentionsInThisNode.length > 0) {
+            let mentionIndex = 0
+            let lastMentionIndex = 0
+            let fragment = document.createDocumentFragment()
+
+            mentionsInThisNode.forEach((mention, i) => {
+              mentionIndex = value.indexOf(mention, lastMentionIndex)
+
+              let mentionWithSpan = document.createElement('span')
+              mentionWithSpan.className = 'mention'
+              mentionWithSpan.id = `mention-${uuidv4()}`
+              mentionWithSpan.textContent = mention
+
+              mentionIndex === 0
+                ? fragment.appendChild(document.createTextNode(''))
+                : fragment.appendChild(document.createTextNode(value.substring(lastMentionIndex, mentionIndex)))
+
+              fragment.appendChild(mentionWithSpan)
+
+              mentionsInThisNode.length - 1 === i
+                ? fragment.appendChild(document.createTextNode(value.substring(mentionIndex + mention.length)))
+                : fragment.appendChild(document.createTextNode(''))
+
+              lastMentionIndex = mentionIndex + mention.length - 1
+            })
+            childNodesList[i].replaceWith(fragment)
+            i = i + 3 * mentionsInThisNode.length
+          } else i++
+        } else {
+          if (!(node.nodeName === 'SPAN' && node.className === 'mention') && node.childNodes.length > 0) depthFirstSearchAndMentionAnalysis(node.childNodes)
```

Comment:
Beware. Should Tracim ever migrate to XHTML, nodeName would become lower case. I'd therefore suggest the following to handle both cases:

```suggestion
          if (!(node.nodeName.toLowerCase() === 'span' && node.className === 'mention') && node.childNodes.length > 0) depthFirstSearchAndMentionAnalysis(node.childNodes)
```
You can also leave out `&& node.childNodes.length > 0`, your code should handle this case (that should not happen anyway) nicely.